src/Analyzers/Analyzers/Serializability/ViewModelSerializabilityAnalyzer.cs

```diff
@@ -0,0 +1,166 @@
+        private static readonly LocalizableResourceString uninstantiableTypeDescription = new LocalizableResourceString(nameof(Resources.Serializability_UninstantiableType_Description), Resources.ResourceManager, typeof(Resources));
+        private const string dotvvmViewModelInterfaceMetadataName =  "DotVVM.Framework.ViewModel.IDotvvmViewModel";
+        private const string dotvvmBindAttributeMetadataName = "DotVVM.Framework.ViewModel.BindAttribute";
+
+        public static DiagnosticDescriptor UseSerializablePropertiesRule = new DiagnosticDescriptor(
+            DotvvmDiagnosticIds.UseSerializablePropertiesInViewModelRuleId,
+            nonSerializableTypeTitle,
+            nonSerializableTypeMessage,
+            DiagnosticCategory.Serializability,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            nonSerializableTypeDescription);
+
+        public static DiagnosticDescriptor DoNotUseFieldsRule = new DiagnosticDescriptor(
+            DotvvmDiagnosticIds.DoNotUseFieldsInViewModelRuleId,
+            doNotUseFieldsTitle,
+            doNotUseFieldsMessage,
+            DiagnosticCategory.Serializability,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            doNotUseFieldsDescription);
+
+        public static DiagnosticDescriptor DoNotUseUninstantiablePropertiesRule = new DiagnosticDescriptor(
+            DotvvmDiagnosticIds.DoNotUseUninstantiablePropertiesInViewModelRuleId,
+            uninstantiableTypeTitle,
+            uninstantiableTypeMessage,
+            DiagnosticCategory.Serializability,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            uninstantiableTypeDescription);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(
+                UseSerializablePropertiesRule,
+                DoNotUseFieldsRule,
+                DoNotUseUninstantiablePropertiesRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.EnableConcurrentExecution();
+            context.RegisterSemanticModelAction(AnalyzeSyntaxTree);
+        }
+
+        private static void AnalyzeSyntaxTree(SemanticModelAnalysisContext context)
+        {
+            var semanticModel = context.SemanticModel;
+            var syntaxTree = context.SemanticModel.SyntaxTree;
+            var viewModelInterface = semanticModel.Compilation.GetTypeByMetadataName(dotvvmViewModelInterfaceMetadataName);
+            var bindAttribute = semanticModel.Compilation.GetTypeByMetadataName(dotvvmBindAttributeMetadataName);
+
+            // Check all classes
+            foreach (var classDeclaration in syntaxTree.GetRoot().DescendantNodes().OfType<ClassDeclarationSyntax>())
```

Comment:
Could we verify whether there is some API that we just add a listener and catch all events for changes of Class Declaration?